digitaltwin_client/src/dt_interface_client.c

```diff
@@ -1465,12 +1504,13 @@ static DIGITALTWIN_CLIENT_RESULT DT_InterfaceClient_SetAsyncResponseProperties(I
 
 // DigitalTwin_InterfaceClient_UpdateAsyncCommandStatusAsync is invoked by application when an async DigitalTwin command is executing, updating
 // DigitalTwin server with status information provided by caller in a specially formatted telemetry message.
-DIGITALTWIN_CLIENT_RESULT DigitalTwin_InterfaceClient_UpdateAsyncCommandStatusAsync(DIGITALTWIN_INTERFACE_CLIENT_HANDLE dtInterfaceClientHandle, const DIGITALTWIN_CLIENT_ASYNC_COMMAND_UPDATE* dtClientAsyncCommandUpdate)
+DIGITALTWIN_CLIENT_RESULT DigitalTwin_InterfaceClient_UpdateAsyncCommandStatusAsync(DIGITALTWIN_INTERFACE_CLIENT_HANDLE dtInterfaceClientHandle, const DIGITALTWIN_CLIENT_ASYNC_COMMAND_UPDATE* dtClientAsyncCommandUpdate, DIGITALTWIN_CLIENT_UPDATE_ASYNC_COMMAND_CALLBACK dtUpdateAsyncCommandCallback, void* userContextCallback)
 {
     DT_INTERFACE_CLIENT* dtInterfaceClient = (DT_INTERFACE_CLIENT*)dtInterfaceClientHandle;
     DIGITALTWIN_CLIENT_RESULT result;
     STRING_HANDLE jsonToSend = NULL;
     IOTHUB_MESSAGE_HANDLE telemetryMessageHandle = NULL;
+    DT_INTERFACE_SEND_TELEMETRY_CALLBACK_CONTEXT* sendTelemetryCallbackContext = NULL;
```

Comment:
>DT_INTERFACE_SEND_TELEMETRY_CALLBACK_CONTEXT [](start = 4, length = 44)

I notice that we are repurposing the DT_INTERFACE_SEND_TELEMETRY_CALLBACK_CONTEXT structure here for both async commands and sending telemetry. I understand the logical structure is the same, but I would rather typedef the DT_INTERFACE_SEND_TELEMETRY_CALLBACK_CONTEXT to something like DT_INTERFACE_ASYNC_COMMAND_CALLBACK_CONTEXT. This way it is apparent to SDK devs that it isn't in error.

Reply:
DT_INTERFACE_SEND_TELEMETRY_CALLBACK_CONTEXT is an internal only structure in this file.  If you look at API shape for app developer, they are different function callbacks exactly so that it's clear to SDK which to use.

I'd noodled over naming on this and the issue is at end of day this is a send telemetry req at protocol level.  I have to have the structure be the same (without much bigger changes, including to dt_client_core.c) because it blindly shoots back telemetry ACK's via DT_InterfaceClient_ProcessTelemetryCallback.  The intf client then handles de-muxing.

---
In reply to: [322957430](https://github.com/Azure/azure-iot-sdk-c/pull/1169#discussion_r322957430) [](ancestors = 322957430)